File: golearning/03_control_flow/control_flow.go
// ============================================================
// 第3章 · 流程控制
// ============================================================
// 运行: go run ./03_control_flow
// 学习要点:
//  1. if：条件不加括号、{ 不能另起一行；还能带初始化语句
//  2. for 是唯一的循环关键字（没有 while / do-while）
//  3. range 遍历：切片/映射/字符串（按 rune 不拆中文）/通道
//  4. switch：默认自动 break、case 可多值可表达式、type switch
//  5. 标签 label：break/continue 直接作用于外层循环
//
// ============================================================
package main

import "fmt"

func main() {
	// ==== 1. if / else if / else ====
	score := 85
	// 与 C/Java 区别：条件不用括号；左花括号必须和 if 同行（gofmt 强制）
	if score >= 90 {
		fmt.Println("优秀")
	} else if score >= 80 {
		fmt.Println("良好")
	} else {
		fmt.Println("继续加油")
	}

	// 特色：if 可带一条初始化语句，分号后接条件
	// 初始化的变量作用域仅限整个 if/else 链——避免污染外部作用域
	if n := score / 10; n >= 9 {
		fmt.Println("档位:", n)
	} else {
		fmt.Println("十位档:", n) // else 分支里 n 仍可见
	}
	// fmt.Println(n) // 编译错误：n 已超出作用域

	// ==== 2. for 的三种形态 ====
	// 形态1：经典三段式（等同 C 的 for）
	for i := 0; i < 3; i++ {
		fmt.Print(i, " ")
	}
	fmt.Println()

	// 形态2：只留条件 = 其他语言的 while（Go 没有 while 关键字！）
	count := 3
	for count > 0 {
		fmt.Print(count, " ")
		count--
	}
	fmt.Println()

	// 形态3：无限循环 = while(true)，靠 break 跳出
	sum := 0
	for {
		sum++
		if sum >= 5 {
			break
		}
	}
	fmt.Println("sum =", sum)

	// 没有 do-while；要模拟就 for { ...; if !cond { break } }

	// ==== 3. range 遍历 ====
	nums := []int{10, 20, 30} // 切片，第5章详讲
	// 返回 (下标, 值)：像 Python 的 enumerate(nums)
	for idx, val := range nums {
		fmt.Printf("nums[%d]=%d ", idx, val)
	}
	fmt.Println()

	for _, val := range nums { // _ 丢弃不需要的下标（Go 不允许存在未使用变量）
		fmt.Print(val, " ")
	}
	fmt.Println()

	// range 字符串按【rune(码点)】迭代而不是字节——中文不会被拆碎！
	// C 遍历 char* 是按字节；Python3 的 str 也是按码点，这里和 Python 一致
	for i, r := range "你好" {
		fmt.Printf("(%d:%c) ", i, r) // 注意 i 是字节偏移：0 和 3
	}
	fmt.Println()

	// range map：顺序故意随机化，防止你依赖遍历顺序
	ages := map[string]int{"张三": 20, "李四": 25}
	for name, age := range ages {
		fmt.Printf("%s=%d ", name, age)
	}
	fmt.Println()

	// ==== 4. switch ====
	// 最大区别：每个 case 自带 break！不像 C 会贯穿到下一个 case
	day := 6
	switch day {
	case 1, 2, 3, 4, 5: // 多个值逗号分隔
		fmt.Println("工作日")
	case 6, 7:
		fmt.Println("周末")
	default:
		fmt.Println("非法日期")
	}
	// 真想要贯穿行为：显式写 fallthrough（很少用到）

	// 无表达式的 switch = 更清爽的 if-else 链（神似 Python 3.10 的 match-case）
	switch h := 21; {
	case h < 12:
		fmt.Println("上午")
	case h < 18:
		fmt.Println("下午")
	default:
		fmt.Println("晚上")
	}

	// case 可以是任意非常量表达式（C 要求编译期常量，Python 的 case 要常量模式）
	switch {
	case score > 90 && day < 6:
		fmt.Println("学霸的工作日")
	default:
		fmt.Println("普通的一天")
	}

	// type switch：判断接口里的动态类型，第8章接口部分实战

	// ==== 5. break/continue 与标签 ====
outer: // 定义标签
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			if j == 1 {
				continue outer // 直接跳过外层本轮——C/Python 的 continue 做不到
			}
			if i == 2 {
				break outer // 直接跳出外层——比 C 的标志位写法优雅
			}
			fmt.Printf("(%d,%d) ", i, j)
		}
	}
	fmt.Println()

	/*
	 * ---- 一句话总结 ----
	 * 一个 for 打天下；range 是遍历瑞士军刀；
	 * switch 默认不贯穿、无表达式更好用；label 解决嵌套循环跳出难题。
	 */
}

File: javalearning/src/03_control_flow/ControlFlow.java
// ============================================================
// 第3章 · 流程控制
// ============================================================
// 运行: java src/03_control_flow/ControlFlow.java
// 学习要点:
//   1. if/else（条件必须是 boolean，对比 C）
//   2. switch 传统写法 vs 箭头表达式
//   3. while / do-while
//   4. for 与增强 for
//   5. break/continue 与标签
// ============================================================

public class ControlFlow {

    public static void main(String[] args) {
        // ==== 1. if / else if / else ====
        // 语法和 C 完全一样，但条件必须是 boolean 类型
        int score = 85;
        if (score >= 90) {
            System.out.println("优秀");
        } else if (score >= 60) {
            System.out.println("及格");
        } else {
            System.out.println("不及格");
        }
        // int x = 1;
        // if (x) {}      // C 能编译，Java 编译错误！必须写 if (x != 0)

        // 三元运算符也和 C 一样：条件 ? 值1 : 值2
        String level = score >= 60 ? "及格" : "不及格";
        System.out.println("三元运算符: " + level);

        // ==== 2. switch 传统写法 ====
        int day = 6;
        switch (day) {
            case 1:
                System.out.println("周一");
                break; // 忘了 break 会"穿透"到下一个 case！经典坑（C 同款）
            case 6:
            case 7:
                System.out.println("周末");
                break;
            default:
                System.out.println("工作日");
                break;
        }

        // ==== 3. switch 箭头表达式（Java 14+）====
        // 不用写 break、不会穿透，还能直接返回值——有点像 Python 的 match-case
        String dayName = switch (day) {
            case 1 -> "周一";
            case 2, 3, 4, 5 -> "工作日"; // 多值合并用逗号
            case 6, 7 -> "周末";
            default -> "未知";
        };
        System.out.println("switch 表达式: " + dayName);

        // ==== 4. while 和 do-while ====
        int count = 0;
        while (count < 3) {           // 先判断后执行，可能一次都不执行
            System.out.print(count + " ");
            count++;
        }
        System.out.println();

        count = 10;
        do {                           // 先执行后判断，至少执行一次（C 同款）
            System.out.println("do-while 至少跑一次, count = " + count);
            count++;
        } while (count < 5);

        // ==== 5. for 循环 ====
        for (int i = 0; i < 5; i++) {  // 初始化; 条件; 更新，和 C 一模一样
            System.out.print(i + " ");
        }
        System.out.println();

        // 增强 for（for-each）：遍历数组/集合，类似 Python 的 for x in items:
        int[] nums = {10, 20, 30};
        for (int n : nums) {           // 读作 "for each int n in nums"
            System.out.print(n + " ");
        }
        System.out.println("\n");
        // 局限：增强 for 拿不到下标、遍历中不能增删元素，需要下标就用普通 for

        // ==== 6. break / continue ====
        for (int i = 0; i < 10; i++) {
            if (i == 3) continue;  // 跳过本次循环（Python 的 continue 相同）
            if (i == 6) break;     // 跳出整个循环（Python 的 break 相同）
            System.out.print(i + " ");
        }
        System.out.println();

        // 标签：跳出到外层循环（Java 特色，类似受限版 goto；Python 没有这机制）
        outer:                       // 标签名 + 冒号
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (i * j == 2) break outer; // 直接跳出两层
                System.out.print("(" + i + "," + j + ") ");
            }
        }
        System.out.println();

        // ==== 7. 综合练习：九九乘法表 ====
        System.out.println("=== 九九乘法表 ===");
        for (int i = 1; i <= 9; i++) {
            for (int j = 1; j <= i; j++) {
                System.out.printf("%d×%d=%-4d", j, i, i * j); // %-4d 左对齐占4位，printf 风格同 C
            }
            System.out.println();
        }

        /*
         * ---- 一句话总结 ----
         * 流程控制语法基本照搬 C；新增的 switch 箭头表达式更安全；
         * 记住两个 Java 特点：条件必须是 boolean、增强 for 不能拿下标。
         */
    }
}
